### ISSUE TITLE
[Enhancement] Pipeline Parallelism: Production Optimizations for GPipe

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#463]
## Goal: Optimize PipelineParallelModel for Production Workloads

Enhance the current GPipe-style pipeline parallelism implementation with industry-standard optimizations to improve efficiency, reduce memory usage, and balance compute loads across pipeline stages.

Related: PR #393

---

## Current State

PipelineParallelModel implements GPipe-style gradient-based backward pass with:
- ✅ Activation passing between adjacent stages (forward)
- ✅ Gradient communication in reverse direction (backward)
- ✅ Gradient accumulation across stages
- ✅ Parameter synchronization after backward pass

**Limitation**: Basic implementation lacks production optimizations used in PyTorch, Megatron-LM, and DeepSpeed pipeline parallelism.

---

### 1. Model-Specific Layer Partitioning Strategies

**Problem**: Current implementation divides parameters uniformly, which doesn't balance compute load.

**Impact**:
- Uneven layer compute costs → pipeline bubbles (idle time)
- Example: Transformer attention layers (expensive) vs layer norms (cheap)
- Pipeline efficiency loss: 20-30% due to imbalanced stages

**Solution**:
- Implement layer-aware partitioning
- Balance compute cost (FLOPs) across stages, not just parameter count
- Support custom partition strategies via callback/delegate

**Example API**:
```csharp
// Custom partition strategy
var partitioner = new LoadBalancedPartitioner<T>(
    model,
    costEstimator: layer => layer.EstimateFlops(),
    numStages: 4
);
var pipelineModel = new PipelineParallelModel<T, TInput, TOutput>(
    model, config, microBatchSize: 4, partitioner);
```

**References**:
- Megatron-LM layer assignment: https://github.com/NVIDIA/Megatron-LM
- PyTorch Pipe balance algorithm

---

### 2. Micro-Batch Scheduling to Reduce Pipeline Bubbles

**Problem**: Sequential micro-batch processing leaves stages idle during pipeline fill/drain.

**Impact**:
- Pipeline bubble overhead: ~12-25% idle time (GPipe paper)
- Example: 4 stages × 8 micro-batches = ~50% bubble in naive schedule

**Solution**:
- Implement 1F1B (One-Forward-One-Backward) scheduling
- Interleave forward and backward passes to keep stages busy
- Support multiple scheduling strategies (GPipe, PipeDream, etc.)

**Example Schedule (1F1B)**:
```
Stage 0: F0 F1 F2 F3 B0 B1 B2 B3
Stage 1:    F0 F1 F2 B0 B1 B2 B3
Stage 2:       F0 F1 B0 B1 B2 B3
Stage 3:          F0 B0 B1 B2 B3
```

**Benefits**:
- Reduces bubble from ~50% to ~12-15%
- Better memory efficiency (doesn't store all forward activations)

**API**:
```csharp
var scheduleStrategy = new OneForwardOneBackwardSchedule(microBatches: 8);
var pipelineModel = new PipelineParallelModel<T, TInput, TOutput>(
    model, config, microBatchSize: 4, schedule: scheduleStrategy);
```

**References**:
- GPipe: https://arxiv.org/abs/1811.06965
- PipeDream: https://arxiv.org/abs/1806.03377

---

### 3. Activation Checkpointing to Reduce Memory Usage

**Problem**: Storing all forward activations for backward pass consumes massive memory.

**Impact**:
- Memory usage: O(num_layers × batch_size × hidden_dim)
- Example: 100 layers × 1024 batch × 4096 hidden = ~1.6 GB per micro-batch
- Limits batch size and model scale

**Solution**:
- Implement activation checkpointing (gradient checkpointing)
- Store only subset of activations, recompute others during backward
- Trade compute for memory (industry-standard technique)

**Strategy**:
```
Store: Every Nth layer activation (checkpoints)
Discard: Intermediate activations
Backward: Recompute discarded activations from checkpoints
```

**Memory Savings**:
- Baseline: O(L) memory for L layers
- With checkpointing: O(√L) memory
- Example: 100 layers → 10× memory reduction

**API**:
```csharp
var checkpointConfig = new ActivationCheckpointConfig
{
    CheckpointEveryNLayers = 10, // Store every 10th layer
    RecomputeStrategy = RecomputeStrategy.Selective
};
var pipelineModel = new PipelineParallelModel<T, TInput, TOutput>(
    model, config, microBatchSize: 4, checkpointing: checkpointConfig);
```

**References**:
- Gradient checkpointing paper: https://arxiv.org/abs/1604.06174
- Megatron-LM checkpointing implementation

---

## Implementation Priority

1. **High Priority**: Micro-batch scheduling (biggest impact on efficiency)
2. **Medium Priority**: Layer partitioning (model-specific, but important for large models)
3. **Medium Priority**: Activation checkpointing (critical for memory-constrained scenarios)

---

## Success Criteria

- ✅ Pipeline bubble overhead reduced from 50% to <15% (via 1F1B scheduling)
- ✅ Memory usage reduced by 5-10× (via activation checkpointing)
- ✅ Balanced compute load across stages (via layer partitioning)
- ✅ API matches PyTorch/Megatron-LM patterns for familiarity

---

## References

- GPipe paper: https://arxiv.org/abs/1811.06965
- PipeDream paper: https://arxiv.org/abs/1806.03377  
- Megatron-LM: https://github.com/NVIDIA/Megatron-LM
- PyTorch Pipe: https://pytorch.org/docs/stable/pipeline.html
- Activation checkpointing: https://arxiv.org/abs/1604.06174
